### PR TITLE
ECL-405 - Other - Unincorporated Association Entity Type

### PR DESCRIPTION
**Pull request by nemanjatomanovic**

### CODE REVIEW

File: app/uk/gov/hmrc/economiccrimelevyregistration/services/RegistrationValidationService.scala [hunk 1]
```diff
@@ -395,4 +396,22 @@ class RegistrationValidationService @Inject() (clock: Clock, schemaValidator: Sc
...
+
+    val validatedCtUtr =
+      validateOptExists(otherEntityJourneyData.isCtUtrPresent, "Presence of corporation Tax Unique Taxpayer Reference")
+        .map(_ => Right(registration))
+
+    if (validatedCtUtr.isValid && otherEntityJourneyData.isCtUtrPresent.get) {
+      validateOptExists(otherEntityJourneyData.ctUtr, "Corporation Tax Unique Taxpayer Reference")
+        .map(_ => Right(registration))
+    } else {
+      validatedCtUtr
```
[the remark this answers is not in the document]
Comment (peter-hazell) [reply app/uk/gov/hmrc/economiccrimelevyregistration/services/RegistrationValidationService.scala:413]: This way follows similar pattern as elsewhere in the validation service

File: test-common/uk/gov/hmrc/economiccrimelevyregistration/EclTestData.scala [hunk 2]
```diff
@@ -81,6 +81,10 @@ final case class ValidCharityRegistration(
   registration: Registration
 )
 
+final case class UnincorporatedAssociationRegistration(
```
Comment (peter-hazell) [inline test-common/uk/gov/hmrc/economiccrimelevyregistration/EclTestData.scala:84]:
```suggestion
final case class ValidUnincorporatedAssociationRegistration(
```

File: app/uk/gov/hmrc/economiccrimelevyregistration/services/RegistrationValidationService.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyregistration/services/RegistrationValidationService.scala:413]: Ah, I was looking for this one, thanks 😄

File: test/uk/gov/hmrc/economiccrimelevyregistration/services/RegistrationValidationServiceSpec.scala [hunk 3]
```diff
@@ -427,5 +427,40 @@ class RegistrationValidationServiceSpec extends SpecBase {
...
+          optOtherEntityJourneyData = Some(otherEntityJourneyData)
+        )
+        val expectedErrors                               = Seq(
+          DataValidationError(DataMissing, "Corporation Tax Unique Taxpayer Reference is missing")
+        )
+
+        val result = service.validateRegistration(invalidUnincorporatedAssociationRegistration)
+        result.isValid shouldBe false
+        result.leftMap(nec => nec.toList should contain theSameElementsAs expectedErrors)
+    }
```
Comment (peter-hazell) [inline test/uk/gov/hmrc/economiccrimelevyregistration/services/RegistrationValidationServiceSpec.scala:464]: what about a test for if the unincorporated association is valid?

File: test/uk/gov/hmrc/economiccrimelevyregistration/services/RegistrationValidationServiceSpec.scala [hunk 4]
```diff
@@ -427,5 +427,54 @@ class RegistrationValidationServiceSpec extends SpecBase {
...
+
+    "return valid result without errors" in {
```
Comment (peter-hazell) [inline test/uk/gov/hmrc/economiccrimelevyregistration/services/RegistrationValidationServiceSpec.scala:466]:
```suggestion
    "return the registration if the registration for an unincorporated association is valid" in {
```

just to keep the description more specific as there are a lot of tests in this spec

Comment (nemanjatomanovic) [reply test/uk/gov/hmrc/economiccrimelevyregistration/services/RegistrationValidationServiceSpec.scala]: Done :)

**Review by peter-hazell — APPROVED** [review]

**Pull request merged**